cmd/webhook-server/main.go

```diff
@@ -0,0 +1,113 @@
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"crypto/tls"
+	"fmt"
+	"github.com/coreos/pkg/capnslog"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	rookclient "github.com/rook/rook/pkg/client/clientset/versioned"
+	"github.com/rook/rook/pkg/clusterd"
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"net/http"
+	"path/filepath"
+)
+
+const (
+	tlsDir      = "/etc/webhook"
+	tlsCertFile = "tls.crt"
+	tlsKeyFile  = "tls.key"
+)
+
+var (
+	cephclusterResource = metav1.GroupVersionResource{Group: "ceph.rook.io", Version: "v1", Resource: "cephclusters"}
+	logger              = capnslog.NewPackageLogger("github.com/rook/rook", "webhookcmd")
+	rookClientSet       rookclient.Interface
+)
+
+func validateResource(req *v1beta1.AdmissionRequest) ([]patchOperation, error) {
+
+	var patches []patchOperation
+
+	//Checks if CephCluster is being updated and rejects the request if the dataDirHostPath is changed from initial value
+	if isCephClusterUpdate(req) {
+
+		//Fetch the existing ceph cluster object
+		found, err := fetchExistingCephCluster(req)
+		if err != nil {
+			return nil, err
+		}
+
+		raw := req.Object.Raw
+		updatedCephCluster := cephv1.CephCluster{}
+		//Deserializing the incoming CephCluster object
+		if _, _, err := universalDeserializer.Decode(raw, nil, &updatedCephCluster); err != nil {
+			return nil, fmt.Errorf("could not deserialize pod object: %v", err)
+		}
+
+		err = validateUpdatedCephCluster(updatedCephCluster, found)
+		if err != nil {
+			return nil, err
+		}
+
+	}
+
+	return patches, nil
+}
+
+func fetchExistingCephCluster(req *v1beta1.AdmissionRequest) (*cephv1.CephCluster, error) {
+	found, err := rookClientSet.CephV1().CephClusters(req.Namespace).Get(req.Name, metav1.GetOptions{})
+	if err != nil {
+		return &cephv1.CephCluster{}, fmt.Errorf("error finding cephcluster object: %v", err)
+	}
+	return found, nil
+}
+
+func isCephClusterUpdate(req *v1beta1.AdmissionRequest) bool {
+	return req.Resource == cephclusterResource && req.Operation == "UPDATE"
+}
+
+//This will start the server
+func StartServer(context *clusterd.Context) {
+	logger.Infof("Starting the webhook")
+	rookClientSet = context.RookClientset
+	certPath := filepath.Join(tlsDir, tlsCertFile)
+	keyPath := filepath.Join(tlsDir, tlsKeyFile)
+
+	keyPair, err := NewTlsKeypairReloader(certPath, keyPath)
+
+	if err != nil {
+		logger.Errorf("error loading certificate: %v", err.Error())
+		return
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle("/validate", admitFuncHandler(validateResource))
+
+	var httpServer *http.Server
+	httpServer = &http.Server{
+		Addr: ":8443",
```

Comment:
8443 used by dashboard, and it is hardcoded here

Reply:
Fixed. Changed port to 8079